### PR TITLE
feat: add custom reducers to estimators [DET-3098]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Introduce custom reducers for estimator trial. From the docstring in the PR:

```
During distributed evaluation, many types of metrics calculated via ``tf.metrics`` or
``tf.keras.metrics`` cannot be aggregated properly from the per-slot final metrics
calculated by each separate Estimator replica. One example is ``tf.metrics.auc``, where
the ROC AUC calculated over predictions and labels from a full dataset cannot be derived
from a set of ROC AUC metrics evaluated over the shards of a dataset. However, with
``make_metric``, the ROC AUC could be calculated in distributed training by calling
``sklearn.metrics.roc_auc_score`` in a custom ``reducer`` function.
```

## Test Plan

Even more manual testing than before, in addition to adding a new unit test and a new parallel test.

## Commentary

The previous version implemented custom metrics via the `tf.keras.metrics.Metric` interface, which hit *very* strange bugs on multi-gpu machines related to XLA trying to place ops on the wrong GPU.  This only happened with custom `tf.keras.metrics.Metric` objects, so a `tf.keras.metrics.Sum()` would work fine, but if you did a no-op inheritance:

```python
class MySum(tf.keras.metrics.Sum):
    pass

```

then the resulting `MySum()` object would trigger the error.

So it turns out that even though the tensorflow documentation for tf.metrics [like this one](https://www.tensorflow.org/api_docs/python/tf/compat/v1/metrics/mean_iou#returns) indicate that tf.metrics are tuples of `(value_tensor, update_op)`, it turns out that tf.metrics are secretly tuples of `(value_op, update_op)`, as indicated by this error message I ran into once:

```
    WARNING: An OutOfRangeError or StopIteration exception is raised by the
    code in FinalOpsHook. This typically means the Ops running by the
    FinalOpsHook have a dependency back to some input source, which should not
    happen. For example, for metrics in tf.estimator.Estimator, all metrics
    functions return two Ops: `value_op` and  `update_op`. Estimator.evaluate
    calls the `update_op` for each batch of the data in input source and, once
    it is exhausted, it call the `value_op` to get the metric values. The
    `value_op` here should have dependency back to variables reading only,
    rather than reading another batch from input. Otherwise, the `value_op`,
    executed by `FinalOpsHook`, triggers another data reading, which ends
    OutOfRangeError/StopIteration. Please fix that.
```

This is a big deal for our API because a `value_tensor` implies that after every `update_op`, the `value_tensor` has the correct value.  But we needed a hook for when to do the allgather, and a `value_op` supports that concept perfectly.

So this current implementation uses the tf.metrics API, which is simpler, more "estimator-y", and it also skirts the mysterious XLA issue.

### CODE REVIEW

File: harness/determined/estimator/_reducer.py [hunk 1]
```diff
@@ -117,52 +117,38 @@ def cross_slot_reduce(self, per_slot_metrics: List[List[Any]]) -> Any:
...
+    context: estimator.EstimatorExperimentalContext,
+    metric: Any,
+    reducer: MetricReducer,
+    numpy_dtype: Any,
+) -> Tuple[tf.Operation, tf.Operation]:
     """
-    _DistributedMetric is a tf.keras Metric which should be compatible with tf.estimators and
-    which uses a MetricReducer to calculate the metric.
+    _distributed_metric returns a tf.metrics-style tuple of (value_op, update_op).  The value_op is
+    apparently read once after all evaluation is completed, which is where we do the allgather and
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:128]: non-blocking: remove "apparently"

File: harness/determined/estimator/_reducer.py [hunk 2]
```diff
@@ -117,52 +117,38 @@ def cross_slot_reduce(self, per_slot_metrics: List[List[Any]]) -> Any:
...
-        allgathered = self._det_context.allgather_metrics(self._det_last_accumulate)
-        result = self._det_reducer.cross_slot_reduce(allgathered)
-        return np.array(result).astype(self._det_np_dtype)
-
-    def result(self) -> tf.Operation:
-        return self._det_result_op
+    last_accumulate = None  # type: Any
+
+    def py_update(metric: Any) -> None:
+        nonlocal last_accumulate
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:139]: question: why do we have this `nonlocal` here? If we need to maintain state can we just make this a class?

**Comment by rb-determined-ai** [comment]
@aaron276h and I spoke offline, and since the two comments are an internal docstring and a coding style discussion, and since this PR is a customer ask but it is holding up the release, we will address the comments in a followup PR later today.
